core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2344,6 +2344,19 @@ sealed trait ZIO[-R, +E, +A]
   final def when(p: => Boolean)(implicit trace: Trace): ZIO[R, E, Option[A]] =
     ZIO.when(p)(self)
 
+  /**
+   * Executes this workflow when value of the specified `FiberRef` satisfies the
+   * predicate.
+   */
+  final def whenFiberRef[S](ref: => FiberRef[S])(f: S => Boolean)(implicit trace: Trace): ZIO[R, E, Option[A]] =
```

Comment:
Possibly this should be `Option[S]`, or maybe `Option[(A, S)]` if we really want to hold onto the `A`.

The reason is that the `S` will change, so the user won't ever have access to the `S` that produced the decision to execute the effect, unless we provide it to them in the return value.